### PR TITLE
Drill-7813: Excel Reader Crashes on Formulae Which Return Strings

### PR DESCRIPTION
**Pull request by cgivre**
# [DRILL-7813](https://issues.apache.org/jira/browse/DRILL-7813): Excel Reader Crashes on Formulae Which Return Strings

## Description

When querying Excel files, if there is a column which had a formula that returned text, the reader would throw a `NumberFormatException` when trying to parse the result. This minor bug fix checks the type of the result to see whether if it is numeric or not, and maps it to the correct column type. 

## Documentation
No user visible changes other than no errors when querying. :-)

## Testing
Added two unit tests to test for text and numeric formulaic columns.

### CODE REVIEW

**Review by dzamo — APPROVED** [review]
My notes

- The new test code covers both the text-valued formula and number-valued formula cases.  
- I looked for some simple, fast alternatives to the exception-driven testing of whether a cell's content can be parsed as a double but I found nothing compelling.
- On line 467 of ExcelBatchReader.java `numericValue` is a throw-away.  Possibly a style improvement is not to assign the value of `cell.getNumericCellValue()` to anything, or to assign it to something explicit like `double throwAway = ...`

In ExcelBatchReader.java

**Comment by cgivre** [comment]
> My notes
> 
> * The new test code covers both the text-valued formula and number-valued formula cases.
> * I looked for some simple, fast alternatives to the exception-driven testing of whether a cell's content can be parsed as a double but I found nothing compelling.

I didn't either unfortunately.  Unfortunately, the streaming reader does place some limits on what is possible. 

> * On line 467 of ExcelBatchReader.java `numericValue` is a throw-away.  Possibly a style improvement is not to assign the value of `cell.getNumericCellValue()` to anything, or to assign it to something explicit like `double throwAway = ...`
> 
> In ExcelBatchReader.java

I'll make that change.  Really all we want to do is see whether it CAN be read.  We don't actually use the value at that point.

**Comment by cgivre** [comment]
@dzamo 
Thanks for the review.  I made the change you requested.  Could leave a `+1` if you're ok with the changes?
Thx!

**Comment by cgivre** [comment]
Commits have been squashed.

File: contrib/format-excel/src/main/java/org/apache/drill/exec/store/excel/ExcelBatchReader.java [hunk 1]
```diff
@@ -459,7 +459,18 @@ private void populateColumnArray(Cell cell, int colPosition) {
...
       // Case if the column is a date or time
       addColumnToArray(rowWriter, excelFieldNames.get(colPosition), MinorType.TIMESTAMP, false);
-    } else if (cellType == CellType.NUMERIC || cellType == CellType.FORMULA || cellType == CellType.BLANK || cellType == CellType._NONE) {
+    } else if (cellType == CellType.FORMULA) {
+      // Cells with formulae can return either strings or numbers.  Unfortunately, there is no way to
+      // determine what he result of the formula is, so we check to see if the result CAN be converted to a
+      // number.  If yes, we assume the cell is numeric.  If not, we catch the error and write the results as strings
+      try {
+        // This tests to see whether the cell can be read as a number. The return value is not needed here.
+        cell.getNumericCellValue();
```
Comment (vvysotskyi) [inline contrib/format-excel/src/main/java/org/apache/drill/exec/store/excel/ExcelBatchReader.java:468]: Can we use here something like `FormulaEvaluator` to evaluate the result of the formula before obtaining the cell type? If I understand correctly, `FormulaEvaluator` replaces formula with its value, so cell type should be correct after that

**Comment by cgivre** [comment]
@vvysotskyi 
I actually figured out another method for doing this that seems to work.

**Review by vvysotskyi — APPROVED** [review]
@cgivre, thanks for making changes, +1

**Comment by cgivre** [comment]
Thanks @vvysotskyi for the review.  Commits squashed and I'll merge this tomorrow.

**Pull request merged**